### ISSUE TITLE
Hive Transform/Map-Reduce Syntax parsing error and TestRig/antlr4test-maven-plugin discrepancy

### ISSUE THREAD

**Issue by oldsong** [antlr/grammars-v4#1817]
Current Hive grammar cannot handle some Transform/Map-Reduce statements like examples in:
[https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Transform#LanguageManualTransform-TRANSFORMExamples](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Transform#LanguageManualTransform-TRANSFORMExamples)

The root cause of the error: `rowFormat`,  `recordReader` and `recordWriter` in Transform statements are all optional, accordingly in the original v3 grammar of Hive source tree each of these rules contains an empty alternative, while in our v4 grammar these empty alternatives are missing. The fix is straightforward, just change all `rowFormat `reference to `rowFormat?`, etc. The fixed grammar can parse all the examples in Hive website above using TestRig.

However some of the examples cannot pass the mvn test. For example:
```
FROM (
  FROM pv_users
  MAP pv_users.userid, pv_users.date
  USING 'map_script'
  AS dt, uid
  CLUSTER BY dt) map_output
INSERT OVERWRITE TABLE pv_users_reduced
  REDUCE map_output.dt, map_output.uid
  USING 'reduce_script'
  AS date, count
```
will trigger error below:
```
line 3:32 no viable alternative at input 'pv_users.date'
line 10:5 mismatched input 'date' expecting {'ASC', 'DESC', 'NULLS', 'LAST', 'OVERWRITE', 'PARTITIONS', 'TABLES', 'COLUMNS', 'INDEX', 'INDEXES', 'REBUILD', 'FUNCTIONS', 'SHOW', 'MSCK', 'REPAIR', 'DIRECTORY', 'CLUSTER', 'DISTRIBUTE', 'SORT', 'LOAD', 'EXPORT', 'REPLICATION', 'METADATA', 'DATA', 'INPATH', 'CHANGE', 'FIRST', 'AFTER', 'RENAME', 'COMMENT', 'TINYINT', 'DATETIME', 'ZONE', 'STRING', 'STRUCT', 'UNIONTYPE', 'PARTITIONED', 'CLUSTERED', 'SORTED', 'BUCKETS', 'FORMAT', 'DELIMITED', 'FIELDS', 'TERMINATED', 'ESCAPED', 'COLLECTION', 'ITEMS', 'KEYS', '$KEY$', 'KILL', 'LINES', 'STORED', 'FILEFORMAT', 'INPUTFORMAT', 'OUTPUTFORMAT', 'INPUTDRIVER', 'OUTPUTDRIVER', 'ENABLE', 'DISABLE', 'LOCATION', 'BUCKET', 'ADD', 'REPLACE', 'TEMPORARY', 'FILE', 'JAR', 'EXPLAIN', 'FORMATTED', 'DEPENDENCY', 'LOGICAL', 'SERDE', 'DEFERRED', 'SERDEPROPERTIES', 'DBPROPERTIES', 'LIMIT', 'OFFSET', 'UNSET', 'TBLPROPERTIES', 'IDXPROPERTIES', '$VALUE$', '$ELEM$', 'DEFINED', 'MAPJOIN', 'STREAMTABLE', 'CLUSTERSTATUS', 'UTC', 'UTC_TMESTAMP', 'LONG', 'PLUS', 'VIEW', 'VIEWS', 'DATABASES', 'MATERIALIZED', 'SCHEMA', 'SCHEMAS', 'SSL', 'UNDO', 'LOCK', 'LOCKS', 'UNLOCK', 'SHARED', 'EXCLUSIVE', 'UNSIGNED', 'WHILE', 'READ', 'PURGE', 'ANALYZE', 'BEFORE', 'CONTINUE', 'RECORDREADER', 'RECORDWRITER', 'SEMI', 'TOUCH', 'ARCHIVE', 'UNARCHIVE', 'COMPUTE', 'STATISTICS', 'USE', 'OPTION', 'CONCATENATE', 'SHOW_DATABASE', 'RESTRICT', 'CASCADE', 'SKEWED', 'DIRECTORIES', 'SETS', 'NOSCAN', 'ROLE', 'ROLES', 'URI', 'SERVER', 'ADMIN', 'OWNER', 'PRINCIPALS', 'COMPACT', 'COMPACTIONS', 'TRANSACTIONS', 'REWRITE', 'REOPTIMIZATION', 'RELOAD', KW_YEAR, 'QUERY', 'QUARTER', KW_MONTH, KW_WEEK, KW_DAY, 'DAYOFWEEK', KW_HOUR, KW_MINUTE, KW_SECOND, 'TRANSACTION', 'WORK', 'WRITE', 'ISOLATION', 'LEVEL', 'SNAPSHOT', 'AUTOCOMMIT', 'CACHE', 'ENFORCED', 'VALIDATE', 'NOVALIDATE', 'RELY', 'NORELY', 'KEY', 'ABORT', 'MATCHED', 'REPL', 'DUMP', 'STATUS', 'VECTORIZATION', 'SUMMARY', 'OPERATOR', 'EXPRESSION', 'DETAIL', 'WAIT', 'RESOURCE', 'PLAN', 'QUERY_PARALLELISM', 'PLANS', 'ACTIVATE', 'DEFAULT', 'CHECK', 'POOL', 'MOVE', 'DO', 'ALLOC_FRACTION', 'SCHEDULING_POLICY', 'PATH', 'MAPPING', 'WORKLOAD', 'MANAGEMENT', 'ACTIVE', 'UNMANAGED', '(', Identifier}
com.khubla.antlr.antlr4test.AssertErrorsException: found 2 errors, but missing file transform_2.sql.errors
        at com.khubla.antlr.antlr4test.AssertErrorsErrorListener.assertErrors(AssertErrorsErrorListener.java:70)
        at com.khubla.antlr.antlr4test.ScenarioExecutor.testGrammar(ScenarioExecutor.java:180)
...
```

The mvn should have done the generating/compiling work correctly, because I can run TestRig tests against these examples smoothly in target/classes directory.

My fork containing the fix and example statements is at:
[https://github.com/oldsong/grammars-v4/tree/master/sql/hive](https://github.com/oldsong/grammars-v4/tree/master/sql/hive)

**Comment by oldsong** [comment]
It turns that antlr4test-maven-plugin has done the right thing: 'date' is a reserved keyword in Hive grammar KW_DATE, so it cannot be used as identifier. I changed all occurrences of 'date' to 'date1' in the test SQL files, mvn test is happy now.

The reason why TestRig ignores the grammar errors in the original test SQLs is still unknown to me. I compared the tree output of these two tools (TestRig uses 'date', plugin uses 'dxxx') and found that the only difference is 'date' vs. 'dxxx'.
